extensions/wikia/AdEngine/js/slot/floatingMedrec.js

```diff
@@ -87,30 +89,53 @@ define('ext.wikia.adEngine.slot.floatingMedrec', [
 				}
 			}
 
-			// AD_MIX_1 and AD_MIX_1B
-			if (abTest.getGroup('AD_MIX') && abTest.getGroup('AD_MIX').indexOf('AD_MIX_1') === 0) {
-				refreshAdIfPossible();
-			}
+				if (context.opts.adMix1Enabled) {
+					refreshAdIfPossible();
+				}
+
+				if (context.opts.adMix3Enabled && shouldChangeSlot($adSlot.offset().top, 6)) {
+					if (refresh.adVisible) {
+						$adSlot.addClass('hidden');
+						$recirculationRail.show();
+					} else {
+						viewabilityHandler.refreshOnView(slotName, 0);
+						$recirculationRail.hide();
+					}
+
+					refresh.adVisible = !refresh.adVisible;
+				}
+
 		}
 
 		function getDifference(currentAdPos) {
 			return currentAdPos > refresh.refreshAdPos ? currentAdPos - refresh.refreshAdPos : refresh.refreshAdPos - currentAdPos;
 		}
 
 		function refreshAdIfPossible() {
-			var currentAdPos = $adSlot.offset().top,
-				heightScrolled = getDifference(currentAdPos),
-				timeDifference = (new Date()) - refresh.lastRefreshTime;
+			var currentAdPos = $adSlot.offset().top;
 
-			if (heightScrolled > 10 && timeDifference > 10000 && refresh.refreshNumber < 3) {
+			if (shouldChangeSlot(currentAdPos, 3)) {
+				viewabilityHandler.refreshOnView(slotName, 0);
+			}
+		}
+
+		function shouldChangeSlot(currentAdPos, maxChanges) {
+			var heightScrolled = getDifference(currentAdPos),
+				timeDifference = (new Date()) - refresh.lastRefreshTime,
+				result = heightScrolled > 10 && timeDifference > 10000 && refresh.refreshNumber < maxChanges;
+
+			if (result) {
 				refresh.lastRefreshTime = new Date();
 				refresh.refreshAdPos = currentAdPos;
 				refresh.refreshNumber++;
-				viewabilityHandler.refreshOnView(slotName, 0);
 			}
+
+			return result;
 		}
 
 		function handleFloatingMedrec() {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Function has a complexity of 11 which is greater than 10 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)